trackintel/preprocessing/staypoints.py

```diff
@@ -72,27 +84,34 @@ def generate_locations(
             db = DBSCAN(eps=epsilon, min_samples=num_samples, algorithm="ball_tree", metric=distance_metric)
 
         if agg_level == "user":
-            location_id_counter = 0
-            # TODO: change into groupby
-            for user_id_this in ret_stps["user_id"].unique():
-                # Slice staypoints array by user. This is not a copy!
-                user_staypoints = ret_stps[ret_stps["user_id"] == user_id_this]
 
-                if distance_metric == "haversine":
-                    # the input is converted to list of (lat, lon) tuples in radians unit
-                    p = np.array([[radians(g.y), radians(g.x)] for g in user_staypoints.geometry])
-                else:
-                    p = np.array([[g.x, g.y] for g in user_staypoints.geometry])
-                labels = db.fit_predict(p)
+            # we create a new column: 'serial_num' where we assign serial numbers to each user id
+            ret_stps = ret_stps.assign(serial_num=(ret_stps["user_id"]).astype("category").cat.codes)
+            if print_progress:
+                tqdm.pandas(desc="User location generation")
+                ret_stps = ret_stps.groupby("user_id", as_index=False).progress_apply(
+                    _generate_locations_per_user,
+                    geo_col=geo_col,
+                    distance_metric=distance_metric,
+                    db=db,
+                    num_total_stps=len(ret_stps),
+                )
+            else:
+                ret_stps = ret_stps.groupby("user_id", as_index=False).apply(
+                    _generate_locations_per_user,
+                    geo_col=geo_col,
+                    distance_metric=distance_metric,
+                    db=db,
+                    num_total_stps=len(ret_stps),
+                )
+
+            ret_stps = ret_stps.drop(columns={"serial_num"})
 
-                # enforce unique lables across all users without changing noise labels
-                max_label = np.max(labels)
-                labels[labels != -1] = labels[labels != -1] + location_id_counter
-                if max_label > -1:
-                    location_id_counter = location_id_counter + max_label + 1
+            # we remove the holes in the staypoint ids to ensure sequential location ids
+            noise_labels = ret_stps["location_id"] == -1
+            ret_stps = ret_stps.assign(location_id=(ret_stps["location_id"]).astype("category").cat.codes)
+            ret_stps.loc[noise_labels, "location_id"] = -1
```

Comment:
I would suggest change line 112-113 into 
`ret_stps.loc[~noise_labels, "location_id"] = (ret_stps.loc[~noise_labels,"location_id"].astype("category").cat.codes)`
Otherwise the sequential location id will start from 1 (0 is the original -1 noise label, but reassigned to -1 in line 113)

Reply:
This line has been removed in the new version